---
title: "Adaptive potential models"
author: "Bioshifts group"
output:   
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
---

# Load in data

## Load libraries

```{r libs, message=FALSE, warning=FALSE}

# remotes::install_github("bbolker/broom.mixed")

rm(list=ls())

list.of.packages <- c(
  "ggplot2", "data.table", "glmmTMB", "lme4", "lmerTest","mgcv","brms",
  "itsadug","mgcViz","pbapply",
  "dplyr", "tidyr", "tidyverse","parallel","ggstats","optimx","broom.mixed","plotly",
  "partR2","scales","patchwork","effects","psych") 

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]

if(length(new.packages)) install.packages(new.packages)

sapply(list.of.packages, require, character.only = TRUE)

source(here::here("R/vis.gam.data_function.R"))
source(here::here("R/rsqglmm.R"))


```

## Load data

```{r splist}

mydataset <- read.csv(here::here("Data/gen_data_final_fonseca.csv"))

```

## Select only shifts in the same direction of velocity

```{r}

mydataset <- mydataset %>%
  filter(shift_vel_sign == "pospos" | shift_vel_sign == "negneg")
# 
# summary(mydataset$SHIFT[which(mydataset$shift_vel_sign=="pospos")])
# summary(mydataset$SHIFT[which(mydataset$shift_vel_sign=="negneg")])
# 
# summary(mydataset$velocity[which(mydataset$shift_vel_sign=="pospos")])
# summary(mydataset$velocity[which(mydataset$shift_vel_sign=="negneg")])


```

## Initial data description

```{r }
# how many data at each shift gradient
mydataset %>%
  group_by(Type) %>%
  tally

# how many data at each shift gradient / parameter
mydataset %>%
  group_by(Type,Param) %>%
  tally

summary(mydataset$velocity[which(mydataset$shift_vel_sign=="pospos")])
summary(mydataset$velocity[which(mydataset$shift_vel_sign=="negneg")])

summary(mydataset$SHIFT[which(mydataset$shift_vel_sign=="pospos")])
summary(mydataset$SHIFT[which(mydataset$shift_vel_sign=="negneg")])

summary(mydataset$velocity[which(mydataset$SHIFT>0)])
summary(mydataset$velocity[which(mydataset$SHIFT<0)])

```

## Latitude data

```{r message=FALSE, warning=FALSE}

mydatatogo <- mydataset  %>%
  dplyr::filter(Type == "LAT",
                SHIFT != 0, # remove non-significant shifts
                # Nucleotide_diversity > 0 # select only GD values > 0
  ) %>% 
  mutate(
    # Climate velocity
    vel = as.numeric(velocity),
    vel_abs = abs(vel),
    vel_abs_log = log(vel_abs),
    vel_abs_log1p = log1p(vel_abs),
    
    # Shift
    SHIFT = SHIFT, # to deal with zero shift
    SHIFT_abs = abs(SHIFT),
    SHIFT_abs_log = log(SHIFT_abs),
    SHIFT_abs_log1p = log1p(SHIFT_abs),
    
    # SHIFT_cor_raw = SHIFT_cor,
    # SHIFT_cor = scale(SHIFT_cor_raw),
    # SHIFT_cor_abs = abs(SHIFT_cor_raw),
    # SHIFT_cor_abs_log = log(SHIFT_cor_abs),
    # SHIFT_cor_abs_log_scale = scale(SHIFT_abs_log),
    
    # Genetic diversity
    GD = Nucleotide_diversity, 
    GD_sqrt = sqrt(GD),
    GD_log = log(GD),
    GD_log1p = log1p(GD),
    
    # Methods
    Lat = abs(Latitude),
    Lat_band = round(Lat,0),
    ID.area = scale(ID.area),
    DUR = scale(DUR),
    LogExtent = log(Extent),
    START = scale(START),
    Param = factor(Param),
    Group = factor(Group),
    spp = factor(spp)) %>%
  
  dplyr::select(
    # Genetic diversity
    GD, GD_log, GD_log1p, GD_sqrt, TajimasD,
    # Shift
    SHIFT, SHIFT_abs, SHIFT_abs_log, SHIFT_abs_log1p, 
    # SHIFT_cor, SHIFT_cor_abs, SHIFT_cor_abs_log, SHIFT_cor_raw, SHIFT_abs_log_scale,
    # Velocity
    vel, vel_abs, vel_abs_log, vel_abs_log1p, 
    trend.mean,
    # Methods + Taxonomy
    Article_ID, 
    Lat, # latitudinal position where GD was collected
    Lat_band,
    DUR, Nperiodes, LogNtempUnits, NtempUnits, Extent, LogExtent, ContinuousGrain, Quality, PrAb, ExtentF, AreaF,
    Param, Group, spp, Class, Order, Family, Genus, 
    ECO, Uncertainty_Parameter, Uncertainty_Distribution, Grain_size, Data, Article_ID,
    # whereas GD data overlaps with the range shift's study area
    overlap_SA,
    # Distance from GD data to Bioshifts SA
    dist_cent, # distance to the centroid of the SA
    dist_N, # distance to the North edge of the SA (equivalent to the distance to the LE at the North hemisphere)
    dist_S, # distance to the South edge of the SA (equivalent to the distance to the TE at the North hemisphere)
    dist_edge # distance to the closest edge of the SA
  ) 

cont_vars <- c(1:14, 16:24, 39:45)

mydatatogo[,cont_vars] <- lapply(mydatatogo[,cont_vars], as.numeric)
mydatatogo[,-cont_vars] <- lapply(mydatatogo[,-cont_vars], function(x) factor(x, levels = unique(x)))


```

## Filter Classes with at least 5 species per Param

```{r}

n_sps = 10

test <- mydatatogo %>%
  group_by(Class,Param) %>%
  dplyr::summarise(N_spp = length(unique(spp))) %>% # how many species per parameter?
  dplyr::filter(N_spp >= n_sps) # select classes with > n_sps per param

test <- mutate(test, Class_Param = paste(Class, Param))

mydatatogo <- mydatatogo %>%
  mutate(Class_Param = paste(Class, Param)) %>%
  filter(Class_Param %in% test$Class_Param) %>%
  select(-Class_Param)

mydatatogo[,-cont_vars] <- lapply(mydatatogo[,-cont_vars], function(x) factor(x, levels = unique(x)))

```

## Extra fixes

```{r}
# # Fix latitudinal bands
# similar <- function(vec, val, bound = 3, index = F) {
#   close.index <- which(abs(vec - val) <= bound)
#   if (index) return(close.index)
#   return(vec[close.index])
# }
# 
# mydatatogo$Lat_band <- cut(mydatatogo$Lat_band,
#                            breaks=seq(0, 90, by=10),
#                            include.lowest=TRUE)
# 
# table(mydatatogo$Lat_band)
# 
# mydatatogo$Lat_band[which(mydatatogo$Lat_band=="(80,90]")] <- "(60,70]"
# mydatatogo$Lat_band[which(mydatatogo$Lat_band=="(70,80]")] <- "(60,70]"
# 
# mydatatogo$Lat_band <- factor(mydatatogo$Lat_band, 
#                               levels = names(which(table(mydatatogo$Lat_band)>0)))
# table(mydatatogo$Lat_band)

# Fix param levels
mydatatogo$Param <- relevel(mydatatogo$Param, ref = "O") 

# Cut GD
mydatatogo$GD_cut <- cut(mydatatogo$GD_sqrt,
                         breaks=seq(min(mydatatogo$GD_sqrt), max(mydatatogo$GD_sqrt), length.out = 5),
                         include.lowest=TRUE)

```

# Modelling framework

Here we test the potential of genetic diversity (GD) in modulating species responses to climate exposure by shifting their ranges. GD is an important component of biodiversity because it provides adaptive potential (Frankham, 2005), facilitates species persistence (Frankham, 2005; Pereira et al., 2012), and provides clues about the historical demography of species (Avise, 2000). We contrast to hypotheses by which GD modulates species response to climate exposure. The first hypothesis proposes GD allows species to better track climate change. In this scenario, GD would confer species' adaptive potential to rapidly colonize new areas that become suitable from climate change. The second hypothesis states that GD allows species to stay in place as climate change. In this scenario GD increases plasticity or fast evolution for the development of adaptations to tolerate new climate conditions.

## Data

Genetic diversity (GD) data was extracted from Fonseca et al. (2023), describing patterns of intraspecific nucleotide diversity. Nucleotide diversity was calculated from aggregated georeferenced mitochondrial DNA for animals (vertebrates, arachnids, and insects) and chloroplast DNA for plants using `phylogatR` (Pelletier et al., 2021; <https://phylogatr.osc.edu>), an open-source platform available through the Ohio Supercomputer Center (OSC). `phylogatR` retrieves genetic and geographic data by integrating information from the GBIF (<https://www.gbif.org/>), GenBank (<https://> www.ncbi.nlm.nih.gov/genbank/), and the BOLD (<http://www>. boldsystems.org/index.php). To minimize data loss and duplication when downloading data from GenBank, duplicate entries with the same accession number, but different GBIF occurrence numbers are checked for similarities in geographic coordinates. Sequences with high values of divergence were removed (likely resulting from inclusion of a sequence with an incorrect species label), a conservative approach that prevents inflation on the calculation of genetic diversity within species. Nucleotide diversity was calculated as the number of polymorphic sites in pairwise sequence comparison, using a modified version of the equation proposed by Nei and Li (1979) to allow for a comparison of sequences of different lengths (Miraldo et al. 2016). Finally, nucleotide diversity between pairs of sequences were averaged to obtain a global estimation of nucleotide diversity for each species, and associated centroid of georeferenced DNA sequences.

We gather range shift estimates from the BioShifts database (Comte et al. 2020). These shift estimates were collated from multiple published literature sources that used different methodological protocols. Previous studies have shown that methodological aspects play an important role in explaining variability in species range shifts (Lenoir et al. 2020). Beyond range shift data, we also gathered from BioShifts climate velocity estimates following the method from Burrows at al. (2014), the position of a species range where range shifts were estimated (leading edge, centroid or trailing edge), and methodological variables representing the number of temporal units (continuous), grain size (small, moderate or large), data type (abundance-based vs occurrence-based), quality (balanced, raw, resurveyed), and study area extent (log km). We focus on latitudinal shifts across terrestrial and marine organisms, excluding elevational range shifts. We also excluded range shifts at the opposite direction of the climate exposure gradient, which can be related to local extinction processes or because species may be tracking climate axes other then temperature (e.g., humidity).

In order to maximize the match of species when merging range shifts and GD datasets, we harmonize species names using a customized function that borrows from the R packages `rgbif` and `bdc`. Our final dataset includes `r nrow(mydatatogo)` shifts for `r length(unique(mydatatogo$spp))` species from `r length(unique(mydatatogo$Class))` major taxonomy classes (`r unique(mydatatogo$Class)`).

## Statistics

We explore how species range shifts (km/year) (response variable) varies in function of climate change velocity (km/year) and GD (Nucleotide diversity) using a linear mixed-effect (LME) modelling framework. Our goal is to estimate whereas GD increases (or decreases) species range shift in response to climate change velocity. Therefore, our model includes an interaction term of genetic diversity with climate change velocity. To account for methodological differences in range shift estimates, we included the methodological variables from BioShifts described above as fixed-effect terms in our models. As our dataset comprises species across multiple taxonomic classes, with different ecology and sensitivity to climate variability, observations are considered as non-independent from each other. Flight capacity would confer birds greater dispersal potential than plants, ectotherms and short lived like insects should be more sensitive to climate variability than endotherms like fishes. We control for taxonomy and ecological variability by including Class as a random-intercept in our model:

$$
Range Shift = Climate Velocity * Genetid Diversity + Methods + (1|Class)
$$

We model absolute values for range shifts and climate velocity as we expect GD to impact the magnitude of the coupling between range shifts and climate velocity, not the direction of this relationship. As we are modeling positive values, we fitted our LME models using a Gamma family with a log link. In addition to these models, we also fitted Gaussian models using log-transformed values of range shift and climate velocity. We test wheres the modulating effect of GD varies across the multiple range positions. Different levels of exposure are expected at different positions of a species range, thereby leading to influencing ecological dynamics. At the warm edge species leave closer to their upper thermal tolerance limit and thus should be more sensitive to warming. In contrast, a warming at the cold edge may represent colonization opportunity. Because GD is at the intraspecific scale, a single GD value can be matched to multiple estimates of range shift estimates across different positions of the range for the same species. To avoid false duplicate data, we run separate analyses to the different range positions.




# Data description

```{r message=FALSE, warning=FALSE}

# N species
length(unique(mydatatogo$spp))
# 1890

# N shifts
nrow(mydatatogo)
# 4675

# N shifts per Class
mydatatogo %>% group_by(Class) %>% tally
#   Class              n
# 1 Insecta         1751
# 2 Aves            1611
# 3 Magnoliopsida    664
# 4 Polypodiopsida    23
# 5 Arachnida        248
# 6 Liliopsida       143
# 7 Actinopterygii   235

# N shifts per range position
table(mydatatogo$Param)
#    O   LE   TE 
# 2298 2082  295

# N shifts per Class and range position
mydatatogo %>% group_by(Class,Param) %>% tally
#            Class Param    n

#  1 Insecta        O       195
#  2 Insecta        LE     1435
#  3 Insecta        TE      121

#  4 Aves           O      1275
#  5 Aves           LE      260
#  6 Aves           TE       76

#  7 Magnoliopsida  O       496
#  8 Magnoliopsida  LE      100
#  9 Magnoliopsida  TE       68

# 10 Polypodiopsida O        23

# 11 Arachnida      LE      248

# 12 Liliopsida     O       108
# 13 Liliopsida     LE       23
# 14 Liliopsida     TE       12

# 15 Actinopterygii O       201
# 16 Actinopterygii LE       16
# 17 Actinopterygii TE       18

table(mydatatogo$ContinuousGrain)

table(mydatatogo$Quality)

table(mydatatogo$PrAb)

table(mydatatogo$Order)

```

## Histograms

### Shift

```{r message=FALSE, warning=FALSE}

hist(mydatatogo$SHIFT_abs)
hist(mydatatogo$SHIFT_abs_log)

```

### GD

```{r message=FALSE, warning=FALSE}

hist(mydatatogo$GD)
hist(mydatatogo$GD_log)

### GD vs Class
ggplot(mydatatogo, aes(GD_log))+
  geom_density(aes(fill = Class), color = NA, alpha = .3)+
  theme_classic() +
  labs(y = "Density", x = "Nucleotide diversity (log)")

m <- aov(log1p(GD) ~ Class, mydatatogo)
summary(m)
TukeyHSD(m)

```

Genetic diversity differs among classes! The exception was for the classes Aves and Insecta, which mean GD is not different.

### Velocity

```{r message=FALSE, warning=FALSE}

hist(mydatatogo$vel_abs)
hist(mydatatogo$vel_abs_log)

```

We should use SHIFT_abs with a Gamma family or SHIFT_abs_log in a gaussian family.

### Distance of GD data and range shift data

```{r message=FALSE, warning=FALSE}

hist(mydatatogo$dist_cent/1000, main = "Distance from study area centroid", xlab = "Distance in km")
hist(mydatatogo$dist_edge/1000, main = "Distance from the edge of study area", xlab = "Distance in km")

```

## Scatter plots

### SHIFT vs Velocity

```{r message=FALSE, warning=FALSE, fig.width=5, fig.height=5}
ggplot(mydatatogo,
       aes(x=vel_abs, y=SHIFT_abs))+
  scale_x_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$vel_abs_log), 
                           max(mydatatogo$vel_abs_log), 
                           length.out = 5),3)))+
  scale_y_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$SHIFT_abs_log), 
                           max(mydatatogo$SHIFT_abs_log), 
                           length.out = 5),3)))+
  geom_point(alpha = .3, size = 3, shape=21, fill = "black")+
  theme_classic()+ 
  theme(aspect.ratio=1)+
  labs(x = "Climate velocity (abs)\nkm/year", 
       y = "Range shift (abs)\nkm/year") +
  geom_smooth(method = "lm", color = "red")

```

```{r fig.width=10, fig.height=5, message=FALSE, warning=FALSE}
ggplot(mydatatogo,
       aes(x=vel_abs, y=SHIFT_abs))+
  scale_x_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$vel_abs_log), 
                           max(mydatatogo$vel_abs_log), 
                           length.out = 5),3)))+
  scale_y_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$SHIFT_abs_log), 
                           max(mydatatogo$SHIFT_abs_log), 
                           length.out = 5),3)))+
  geom_point(alpha = .3, size = 3, shape=21, fill = "black")+
  theme_classic()+ 
  theme(aspect.ratio=1)+
  labs(x = "Climate velocity (abs)\nkm/year", 
       y = "Range shift (abs)\nkm/year") +
  geom_smooth(method = "lm", color = "red") +
  facet_wrap(.~Param)

```

### SHIFT vs GD

```{r fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

ggplot(mydatatogo,
       aes(x=GD, y=SHIFT_abs,color=vel_abs))+
  scale_y_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$SHIFT_abs_log), 
                           max(mydatatogo$SHIFT_abs_log), 
                           length.out = 6),3)))+
  scale_color_viridis_c(option = "inferno",direction=-1)+
  geom_point(alpha = .5, size = 3)+
  theme_classic()+
  theme(legend.justification = 1,
        legend.position="top",
        aspect.ratio=1)+
  labs(x = "Nucleotide diversity", 
       y = "Range shift (abs)\nkm/year", 
       color = "Climate change velocity (abs)\nkm/year") +
  geom_smooth(method = "lm", color = "red")

```

```{r fig.width=10, fig.height=5, message=FALSE, warning=FALSE}

ggplot(mydatatogo,
       aes(x=GD, y=SHIFT_abs,color=vel_abs))+
  scale_y_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$SHIFT_abs_log), 
                           max(mydatatogo$SHIFT_abs_log), 
                           length.out = 6),3)))+
  scale_color_viridis_c(option = "inferno",direction=-1)+
  geom_point(alpha = .5, size = 3)+
  theme_classic()+
  theme(legend.justification = 1,
        legend.position="top",
        aspect.ratio=1)+
  labs(x = "Nucleotide diversity", 
       y = "Range shift (abs)\nkm/year", 
       color = "Climate change velocity (abs)\nkm/year") +
  geom_smooth(method = "lm", color = "red") +
  facet_wrap(.~Param)

```

```{r fig.height=7, fig.width=9, message=FALSE, warning=FALSE}
ggplot(mydatatogo,
       aes(x=GD, y=SHIFT_abs, color=vel_abs))+
  scale_y_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$SHIFT_abs_log), 
                           max(mydatatogo$SHIFT_abs_log), 
                           length.out = 6),3)))+
  scale_color_viridis_c(option = "inferno",direction=-1)+
  geom_point(alpha = .3, size = 3)+
  theme_classic()+
  theme(legend.justification = 1,
        legend.position="top",
        aspect.ratio=1)+
  labs(x = "Nucleotide diversity", 
       y = "Range shift (abs)\nkm/year", 
       color = "Climate change velocity (abs)\nkm/year") +
  geom_smooth(method = "lm", color = "red") +
  facet_wrap(Class~., scales = "free")

```

### GD vs Velocity

```{r message=FALSE, warning=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=vel_abs, y=GD))+
  geom_point(alpha = .3, size = 3, shape=21, fill = "black")+
  scale_x_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$vel_abs_log), 
                           max(mydatatogo$vel_abs_log), 
                           length.out = 5),3)))+
  theme_classic()+
  theme(aspect.ratio=1)+
  labs(x = "Climate change velocity (abs)\nkm/year", 
       y = "Nucleotide diversity") +
  geom_smooth(method = "lm", color = "red")

```

```{r fig.width=10, fig.height=5, message=FALSE, warning=FALSE}
ggplot(mydatatogo,
       aes(x=vel_abs, y=GD))+
  geom_point(alpha = .3, size = 3, shape=21, fill = "black")+
  scale_x_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$vel_abs_log), 
                           max(mydatatogo$vel_abs_log), 
                           length.out = 5),3)))+
  theme_classic()+
  theme(aspect.ratio=1)+
  labs(x = "Climate change velocity (abs)\nkm/year", 
       y = "Nucleotide diversity") +
  geom_smooth(method = "lm", color = "red") +
  facet_wrap(.~Param)
```

### SHIFT vs TajimasD

```{r fig.width=5, fig.height=5, message=FALSE, warning=FALSE}

ggplot(mydatatogo,
       aes(x=TajimasD, y=SHIFT_abs,color=vel_abs))+
  scale_color_viridis_c(option = "inferno",direction=-1)+
  geom_point(alpha = .5, size = 3)+
  scale_y_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$SHIFT_abs_log), 
                           max(mydatatogo$SHIFT_abs_log), 
                           length.out = 5),3)))+
  theme_classic()+
  theme(legend.justification = 1,
        legend.position="top",
        aspect.ratio=1)+
  labs(x = "Tajima's D", 
       y = "Range shift velocity (abs)\nkm/year", 
       color = "Climate change velocity (abs)\nkm/year") +
  geom_smooth(method = "lm", color = "red")

```

```{r fig.width=10, fig.height=5, message=FALSE, warning=FALSE}
ggplot(mydatatogo,
       aes(x=TajimasD, y=SHIFT_abs_log, color=vel_abs))+
  scale_y_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$SHIFT_abs_log), max(mydatatogo$SHIFT_abs_log), length.out = 6),3)))+
  scale_color_viridis_c(option = "inferno",direction=-1)+
  geom_point(alpha = .5, size = 3)+
  theme_classic()+
  theme(legend.justification = 1,
        legend.position="top",
        aspect.ratio=1)+
  labs(x = "Tajima's D", 
       y = "Range shift velocity (abs)\nkm/year", 
       color = "Climate change velocity (abs)\nkm/year") +
  geom_smooth(method = "lm", color = "red") +
  facet_wrap(.~Param)
```

```{r fig.height=7.5, fig.width=8, message=FALSE, warning=FALSE}

ggplot(mydatatogo,
       aes(x=TajimasD, y=SHIFT_abs_log, color=vel_abs))+
  scale_y_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$SHIFT_abs_log), max(mydatatogo$SHIFT_abs_log), length.out = 6),3)))+
  scale_color_viridis_c(option = "inferno",direction=-1)+
  geom_point(alpha = .5, size = 3)+
  theme_classic()+
  theme(legend.justification = 1,
        legend.position="top",
        aspect.ratio=1)+
  labs(x = "Tajima's D", 
       y = "Range shift velocity (abs)\nkm/year", 
       color = "Climate change velocity (abs)\nkm/year") +
  geom_smooth(method = "lm", color = "red") +
  facet_wrap(Class~., scales = "free")

```

### Tajima's D vs Velocity

```{r message=FALSE, warning=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=vel_abs, y=TajimasD))+
  scale_x_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$vel_abs_log), 
                           max(mydatatogo$vel_abs_log), 
                           length.out = 5),3)))+
  geom_point(alpha = .3, size = 3, shape=21, fill = "black")+
  theme_classic()+
  theme(aspect.ratio=1)+
  labs(x = "Climate change velocity (abs)\nkm/year", 
       y = "Tajima's D") +
  geom_smooth(method = "lm", color = "red")

```

```{r fig.width=10, fig.height=5, message=FALSE, warning=FALSE}
ggplot(mydatatogo,
       aes(x=vel_abs, y=TajimasD))+
  scale_x_continuous(
    trans="log",
    labels=function(x) round(x,3),
    breaks = exp(round(seq(min(mydatatogo$vel_abs_log), 
                           max(mydatatogo$vel_abs_log), 
                           length.out = 5),3)))+
  geom_point(alpha = .3, size = 3, shape=21, fill = "black")+
  theme_classic()+
  theme(aspect.ratio=1)+
  labs(x = "Climate change velocity (abs)\nkm/year", 
       y = "Tajima's D") +
  geom_smooth(method = "lm", color = "red") +
  facet_wrap(.~Param)
```


## Spatial distribution of data

```{r out.height = "700px", out.width='800px', echo=F}

knitr::include_graphics("./images/Map2.pdf")

```

# Models

## Centroid model

We first test if the species range shift varies in function of both climate change velocity, genetic diversity and the interaction between these two variables, while account for methodology and phylogeny.

### Gaussian model

```{r}

# Model with zero GD
gau1_O <- lmer(SHIFT_abs_log ~ vel_abs_log * GD + 
                 # LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                 (1 | Class/Family), 
               data = mydatatogo %>%
                 filter(Param=="O"))
summary(gau1_O)
round(MuMIn::r.squaredGLMM(gau1_O),2)
AIC(gau1_O)

# Model without zero GD
gau2_O <- lmer(SHIFT_abs_log ~ vel_abs_log * GD +
                 # LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                 (1 | Class/Family), 
               data = mydatatogo %>%
                 filter(Param=="O") %>% 
                 filter_all(all_vars(!is.infinite(.))))
summary(gau2_O)
round(MuMIn::r.squaredGLMM(gau2_O),2)
AIC(gau2_O)

```

The interaction between velocity and GD is not significant in the Gaussian model

```{r eval=FALSE, include=FALSE}
# Partition the variance explained in this model to see how much of the variance is due to methods alone...


# cond_partR <- mergeR2(
#   partR2(gau1_O,
#          partvars = "vel_abs_log",
#          R2_type = "conditional",
#          nboot = 100,
#          parallel = TRUE),
#   partR2(gau1_O,
#          partvars = "GD",
#          R2_type = "conditional",
#          nboot = 100,
#          parallel = TRUE),
#   partR2(gau1_O,
#          partvars = "vel_abs_log:GD",
#          R2_type = "conditional",
#          nboot = 100,
#          parallel = TRUE),
#   partR2(gau1_O,
#          partbatch = list(Methods = c("LogNtempUnits","LogExtent","ContinuousGrain","PrAb","Quality")),
#          R2_type = "conditional",
#          nboot = 100,
#          parallel = TRUE))
# 
# marg_partR <- mergeR2(
#   partR2(gau1_O,
#          partvars = "vel_abs_log",
#          nboot = 100,
#          parallel = TRUE),
#   partR2(gau1_O,
#          partvars = "GD",
#          nboot = 100,
#          parallel = TRUE),
#   partR2(gau1_O,
#          partvars = "vel_abs_log:GD",
#          nboot = 100,
#          parallel = TRUE),
#   partR2(gau1_O,
#          partbatch = list(Methods = c("LogNtempUnits","LogExtent","ContinuousGrain","PrAb","Quality")),
#          nboot = 100,
#          parallel = TRUE))
# 
# # Part R2 unique to each predictor and combinations of (fixed effect) predictors
# (forestplot(marg_partR) + forestplot(cond_partR)) + plot_annotation(tag_levels = "A", tag_prefix= "(", tag_suffix = ")")
# 
# # Inclusive R2 defined as the variance explained by a predictor irrespective of covariances with other predictors
# (forestplot(marg_partR, type = "IR2") + forestplot(cond_partR, type = "IR2")) + plot_annotation(tag_levels = "A", tag_prefix= "(", tag_suffix = ")")
# 
# # Structure coefficients (SC), the contribution of a predictor to model prediction irrespective of other predictors (see Yeatts et al., 2017)
# (forestplot(marg_partR, type = "SC") + forestplot(cond_partR, type = "SC")) + plot_annotation(tag_levels = "A", tag_prefix= "(", tag_suffix = ")")

# Methods explain most of the variability in range shifts. However, the effect of ecology (velocity and GD) is still significant...

```

Check if a Gamma model is better than the Gaussian model:

### Gamma model

```{r}

# Model with zero GD
gam1_O <- glmer(SHIFT_abs ~ vel_abs * GD +
                  LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                  (1 | Class),
                family = Gamma(link = "log"),
                data = mydatatogo %>%
                  filter(Param=="O"))
# model fails to converge
# find_optimizer <- allFit(gam1_O, parallel = "multicore", ncpus = 5)
# find_optimizer$bobyqa@optinfo # find information on optimizer
# a model with bobyqa optimizer converges
gam1_O <- update(gam1_O,
                 control = glmerControl(optimizer = "bobyqa"))
summary(gam1_O)
round(MuMIn::r.squaredGLMM(gam1_O),2)
AIC(gam1_O)

# Model without zero GD
gam2_O <- glmer(SHIFT_abs ~ vel_abs * GD + 
                  LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                  (1 | Class), 
                family = Gamma(link = "log"),
                data = mydatatogo %>%
                  filter(Param=="O") %>% 
                  filter_all(all_vars(!is.infinite(.))))
summary(gam2_O)
round(MuMIn::r.squaredGLMM(gam2_O),2)
AIC(gam2_O)

```

Both models show significant effects for the interaction term, with smaller effect strength in the model that exclude zero GD. Gamma models fit the data better relative to Gaussian models (based AIC and R2 values).

### Interpretation of results in light of genetic diversity

```{r message=FALSE, warning=FALSE, fig.width=7, fig.height=5}

# With zero GD
ef_gam <- effects::Effect(c("vel_abs","GD"), gam1_O)
ef_gam <- as.data.frame(ef_gam)
ef_gam$GD <- as.factor(ef_gam$GD)

ggplot(ef_gam, aes(vel_abs, fit, colour=GD, fill=GD))+
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Climate velocity (abs)\nkm/year", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Nucleotide diversity") +
  guides(fill = FALSE)

# Excluding zero GD
ef_gam <- effects::Effect(c("vel_abs","GD"), gam2_O)
ef_gam <- as.data.frame(ef_gam)
ef_gam$GD <- as.factor(ef_gam$GD)

ggplot(ef_gam, aes(vel_abs, fit, colour=GD, fill=GD))+
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Climate velocity (abs)\nkm/year", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Nucleotide diversity") +
  guides(fill = FALSE)

```

As climate velocity increases, low GD species have to move as faster than high GD species to keep up with climate change. This corroborates the hypotheses that GD increases species' ability to stay in place. The fact that low GD species show a positive relationship with climate velocity indicates that climate exposure poses higher pressure on them relative to high GD species. As a consequence, low GD species have to move to track changing climates and stay under their thermal niches, otherwise they would go extinct. As GD increases, the effect of climate exposure on species range shift is relaxed -- species track climate change less closely.

### Interpretation of results in light of climate velocity

```{r message=FALSE, warning=FALSE, fig.width=7, fig.height=5}

# With zero GD
ef_gam <- effects::Effect(c("GD","vel_abs"), gam1_O)
ef_gam <- as.data.frame(ef_gam)
ef_gam$vel_abs <- as.factor(ef_gam$vel_abs)

ggplot(ef_gam, aes(GD, fit, colour=vel_abs, fill=vel_abs)) +
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Genetic diversity", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Climate velocity (abs)\nkm/year") +
  guides(fill = FALSE)

# Excluding zero GD
ef_gam <- effects::Effect(c("GD","vel_abs"), gam2_O)
ef_gam <- as.data.frame(ef_gam)
ef_gam$vel_abs <- as.factor(ef_gam$vel_abs)

ggplot(ef_gam, aes(GD, fit, colour=vel_abs, fill=vel_abs)) +
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Genetic diversity", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Climate velocity (abs)\nkm/year") +
  guides(fill = FALSE)

```

GD increases species' ability to stay in place under increasing climate exposure. However, when exposure is relaxed (slow climate velocity), GD facilitates species movement, presumably through higher adaptive potential that allow to explore of new habitats and niche axes.

Next we test if these mechanisms hold across TE and LE.

## Leading edge model

```{r}

# Model with zero GD
gam1_LE <- glmer(SHIFT_abs ~ vel_abs * GD +
                   LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                   (1 | Class),
                 family = Gamma(link = "log"),
                 data = mydatatogo %>%
                   filter(Param=="LE"))
summary(gam1_LE)
round(MuMIn::r.squaredGLMM(gam1_LE),2)
AIC(gam1_LE)

# Model without zero GD
gam2_LE <- glmer(SHIFT_abs ~ vel_abs * GD + 
                   LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                   # (1|Article_ID) +
                   (1 | Class), 
                 family = Gamma(link = "log"),
                 data = mydatatogo %>%
                   filter(Param=="LE") %>% 
                   filter_all(all_vars(!is.infinite(.))))
summary(gam2_LE)
round(MuMIn::r.squaredGLMM(gam2_LE),2)
AIC(gam2_LE)


```

The interaction term is significant at the leading edge. However, the significance is marginal when excluding zero GD. Note that the marginal R2 of these models are much smaller than the centroid models.

### Interpretation of results in light of genetic diversity

```{r message=FALSE, warning=FALSE, fig.width=7, fig.height=5}

# With zero GD
ef_gam <- effects::Effect(c("vel_abs","GD"), gam1_LE)
ef_gam <- as.data.frame(ef_gam)
ef_gam$GD <- as.factor(ef_gam$GD)

ggplot(ef_gam, aes(vel_abs, fit, colour=GD, fill=GD))+
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Climate velocity (abs)\nkm/year", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Nucleotide diversity") +
  guides(fill = FALSE)

# Excluding zero GD
ef_gam <- effects::Effect(c("vel_abs","GD"), gam2_LE)
ef_gam <- as.data.frame(ef_gam)
ef_gam$GD <- as.factor(ef_gam$GD)

ggplot(ef_gam, aes(vel_abs, fit, colour=GD, fill=GD))+
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Climate velocity (abs)\nkm/year", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Nucleotide diversity") +
  guides(fill = FALSE)

```

As climate velocity increases at the leading edge, low GD species move faster, whereas high GD species move slower. This seemingly contra-intuitive pattern may indicate that when the pressure from climate exposure is relaxed (i.e., climate stability or slow climate velocity), GD facilitates species movement, presumably though higher adaptive potential that may allow exploration of new habitats and niche axes. However, as climate exposure increases high GD helps species to stay in place.

### Interpretation of results in light of climate velocity

```{r message=FALSE, warning=FALSE, fig.width=7, fig.height=5}

# With zero GD
ef_gam <- effects::Effect(c("GD","vel_abs"), gam1_LE)
ef_gam <- as.data.frame(ef_gam)
ef_gam$vel_abs <- as.factor(ef_gam$vel_abs)

ggplot(ef_gam, aes(GD, fit, colour=vel_abs, fill=vel_abs)) +
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Genetic diversity", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Climate velocity (abs)\nkm/year") +
  guides(fill = FALSE)

# Excluding zero GD
ef_gam <- effects::Effect(c("GD","vel_abs"), gam2_LE)
ef_gam <- as.data.frame(ef_gam)
ef_gam$vel_abs <- as.factor(ef_gam$vel_abs)

ggplot(ef_gam, aes(GD, fit, colour=vel_abs, fill=vel_abs)) +
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Genetic diversity", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Climate velocity (abs)\nkm/year") +
  guides(fill = FALSE)

```

GD increases species' ability to stay in place under increasing climate exposure. However, when exposure is relaxed, GD facilitates species movement, presumably though higher adaptive potential that may allow exploration of new habitats and niche axes.

## Trailing edge model

```{r}

# Model with zero GD
gam1_TE <- glmer(SHIFT_abs ~ vel_abs * GD +
                   LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                   (1 | Class),
                 family = Gamma(link = "log"),
                 data = mydatatogo %>%
                   filter(Param=="TE"))
summary(gam1_TE)
round(MuMIn::r.squaredGLMM(gam1_TE),2)
AIC(gam1_TE)

# Model without zero GD
gam2_TE <- glmer(SHIFT_abs ~ vel_abs * GD + 
                   LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                   (1 | Class), 
                 family = Gamma(link = "log"),
                 data = mydatatogo %>%
                   filter(Param=="TE") %>% 
                   filter_all(all_vars(!is.infinite(.))))
summary(gam2_TE)
round(MuMIn::r.squaredGLMM(gam2_TE),2)
AIC(gam2_TE)


```

The interaction term is not significant at the trailing edge.

## All edges in single model

*Cautionary note:* As the genetic data is at the species-level, we have no enough information about genetic diversity variability across multiple positions of a species range. Despite the issues that may arise from the inclusion of false duplicates (same genetic diversity information across multiple range positions of the same species), we fit a model including all three range positions with the aim to test if the mediating effect of GD on climate exposure is different across multiple range positions. However, given the issues described above, we advise caution when interpreting these results. This model is simplistic as it assumes no geographical variability in GD across a species range.

```{r}

# Model with zero GD
gam1 <- glmer(SHIFT_abs ~ scale(vel_abs) * scale(GD) * Param +
                LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                (1 | Class),
              family = Gamma(link = "log"),
              data = mydatatogo)
summary(gam1)
round(MuMIn::r.squaredGLMM(gam1),2)
AIC(gam1)

# Model without zero GD
gam2 <- glmer(SHIFT_abs ~ scale(vel_abs) * scale(GD) * Param +
                LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                (1 | Class), 
              family = Gamma(link = "log"),
              data = mydatatogo %>% 
                filter_all(all_vars(!is.infinite(.))))
summary(gam2)
round(MuMIn::r.squaredGLMM(gam2),2)
AIC(gam2)

```

In both models the interaction term of velocity:GD:Param is not significant. This suggests that the effect of GD and velocity are consistent across species range positions.

### Interpretation of results in light of genetic diversity

```{r message=FALSE, warning=FALSE, fig.width=15, fig.height=5}

# With zero GD
ef_gam <- effects::Effect(c("vel_abs","GD","Param"), gam1)
ef_gam <- as.data.frame(ef_gam)
ef_gam$GD <- as.factor(ef_gam$GD)
ef_gam$Param <- factor(ef_gam$Param,levels = c("TE","O","LE"))

ggplot(ef_gam, aes(vel_abs, fit, colour=GD, fill=GD))+
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Climate velocity (abs)\nkm/year", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Nucleotide diversity") +
  guides(fill = FALSE) +
  facet_wrap(.~Param, scales = "free")

# Excluding zero GD
ef_gam <- effects::Effect(c("vel_abs","GD","Param"), gam2)
ef_gam <- as.data.frame(ef_gam)
ef_gam$GD <- as.factor(ef_gam$GD)
ef_gam$Param <- factor(ef_gam$Param,levels = c("TE","O","LE"))

ggplot(ef_gam, aes(vel_abs, fit, colour=GD, fill=GD))+
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Climate velocity (abs)\nkm/year", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Nucleotide diversity") +
  guides(fill = FALSE) +
  facet_wrap(.~Param, scales = "free")

```

Across all range edges, low GD species move faster as climate velocity increases, whereas high GD species move slower.

### Interpretation of results in light of climate velocity

```{r message=FALSE, warning=FALSE, fig.width=15, fig.height=5}

# With zero GD
ef_gam <- effects::Effect(c("GD","vel_abs","Param"), gam1)
ef_gam <- as.data.frame(ef_gam)
ef_gam$vel_abs <- as.factor(ef_gam$vel_abs)
ef_gam$Param <- factor(ef_gam$Param,levels = c("TE","O","LE"))

ggplot(ef_gam, aes(GD, fit, colour=vel_abs, fill=vel_abs)) +
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Genetic diversity", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Climate velocity (abs)\nkm/year") +
  guides(fill = FALSE) +
  facet_wrap(.~Param, scales = "free")

# Excluding zero GD
ef_gam <- effects::Effect(c("GD","vel_abs","Param"), gam2)
ef_gam <- as.data.frame(ef_gam)
ef_gam$vel_abs <- as.factor(ef_gam$vel_abs)
ef_gam$Param <- factor(ef_gam$Param,levels = c("TE","O","LE"))

ggplot(ef_gam, aes(GD, fit, colour=vel_abs, fill=vel_abs)) +
  geom_line(size=1)+
  geom_ribbon(colour=NA,alpha=0.1,
              aes(ymin=lower,ymax=upper))+
  scale_color_viridis_d(option = "C")+
  scale_fill_viridis_d(option = "C")+
  theme_classic()+ 
  theme(legend.justification = 1,
        aspect.ratio=1)+
  labs(x = "Genetic diversity", 
       y = "Predicted range shift (abs)\nkm/year",
       color = "Climate velocity (abs)\nkm/year") +
  guides(fill = FALSE) +
  facet_wrap(.~Param, scales = "free")

```

Across all positions of a species range, GD increases species' ability to stay in place under increasing climate exposure. However, when exposure is relaxed, GD facilitates species movement, presumably though higher adaptive potential that may allow exploration of new habitats and niche axes.

# Sensitive analyses

Here we perform supplementary analyses to explore the robustness of our results to possible bias that may exist in our data and analyses.

## Outliers

Outliers can have a lot of influence on the results when considering interactions so it would be useful to get a sense of the uncertainty. We adopt a multivariate outliers detection approach based on Mahalanobis Distance (MD). MD calculates the distance of each observation from the central mean. Larger values indicate that a observation is farther from where most of the points cluster. We calculate the MDs using the mahalanobis function from the `psych` package and identify those that fall above the cut-off score for a chi-square with k degrees of freedom (3 for 3 variables):

```{r}

data_outlier <- mydatatogo[,c("SHIFT_abs","vel_abs","GD")]

md <- mahalanobis(data_outlier, 
                  center = colMeans(data_outlier), 
                  cov = cov(data_outlier))
alpha <- .001
cutoff <- (qchisq(p = 1 - alpha, df = 3))
names_outliers_MH <- which(md > cutoff)
excluded_mh <- names_outliers_MH

mydatatogo_sub <- mydatatogo[-excluded_mh, ]

{
  par(mfrow=c(1,2))
  hist(mydatatogo$SHIFT_abs, main="With outliers", xlab = "Range shift")
  hist(mydatatogo_sub$SHIFT_abs, main="Without outliers", xlab = "Range shift")
}
{
  par(mfrow=c(1,2))
  hist(mydatatogo$GD, main="With outliers", xlab = "Genetic diversity")
  hist(mydatatogo_sub$GD, main="Without outliers", xlab = "Genetic diversity")
}
{
  par(mfrow=c(1,2))
  hist(mydatatogo$vel_abs, main="With outliers", xlab = "Climate velocity")
  hist(mydatatogo_sub$vel_abs, main="Without outliers", xlab = "Climate velocity")
}


```

### Rerun analysis - Centroid model

```{r}

# Model with zero GD
gam1_O <- glmer(SHIFT_abs ~ vel_abs * GD +
                  LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                  (1 | Class),
                family = Gamma(link = "log"),
                data = mydatatogo_sub %>%
                  filter(Param=="O"))
summary(gam1_O)
round(MuMIn::r.squaredGLMM(gam1_O),2)
AIC(gam1_O)

# We do not have to fit a model without zero GD because they are removed as outliers

```

Results are consistent when removing outliers!

### Rerun analysis - Leading edge model

```{r}

# Model with zero GD
gam1_LE <- glmer(SHIFT_abs ~ vel_abs * GD +
                   LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                   (1 | Class),
                 family = Gamma(link = "log"),
                 data = mydatatogo_sub %>%
                   filter(Param=="LE"))
summary(gam1_LE)
round(MuMIn::r.squaredGLMM(gam1_LE),2)
AIC(gam1_LE)

# We do not have to fit a model without zero GD because they are removed as outliers

```

At the leading edge, removing outliers make the interaction term and also the effects of velocity and GD insignificant.

### Rerun analysis - Trailing edge model

```{r}

# Model with zero GD
gam1_TE <- glmer(SHIFT_abs ~ vel_abs * GD +
                   LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                   (1 | Class),
                 family = Gamma(link = "log"),
                 data = mydatatogo_sub %>%
                   filter(Param=="TE"))
summary(gam1_TE)
round(MuMIn::r.squaredGLMM(gam1_TE),2)
AIC(gam1_TE)

# We do not have to fit a model without zero GD because they are removed as outliers

```

Results also no longer significant at the trailing edge (just as when outliers are included).

### Rerun analysis - Fitting all edges in single model

```{r}

# Model with zero GD
gam1 <- glmer(SHIFT_abs ~ scale(vel_abs) * scale(GD) * Param +
                LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                (1 | Class),
              family = Gamma(link = "log"),
              data = mydatatogo_sub)
summary(gam1)
round(MuMIn::r.squaredGLMM(gam1),2)
AIC(gam1)

# We do not have to fit a model without zero GD because they are removed as outliers

```

The interaction term of velocity:GD is no longer significant. Yet, the interaction term of velocity:GD:Param is also not significant.

## False duplicates

In order to accounting for the issues that may arise from false duplicate samples for GD (see section Modelling Framework above), we reran analyses by subsetting our complete datasets to include only one range edge shift per species. Doing so, we applied prioritization levels for each range edge. For example, for the model in which the centroid was prioritized, species with range shifts estimated at other range positions (e.g., centroid + leading edge) were excluded. For the remaining species with shifts outside of the centroid, we apply a second level of prioritization, finally keeping one range shift observation per species and per edge. This approach was repeated for all three range edges and prioritization levels resulting in six datasets:

```{r echo=FALSE}

priority_data <- data.frame(Dataset = 1:6,
                            First = c("O","O", "LE", "LE", "TE", "TE"), 
                            Second = c("LE","TE","O","TE","O","LE"), 
                            Third = c("TE","LE","TE","O","LE","O"))

priority_data

```

```{r}

# get edges per species
edge_data <- mydatatogo %>%
  group_by(spp, Param) %>%
  tally %>%
  mutate(n = 1) %>%
  pivot_wider(names_from = Param, values_from = n) 

edge_data[,2:4] <- ifelse(is.na(edge_data[,2:4]), 0, 1) 
edge_data <- edge_data %>%
  rowwise() %>% 
  mutate(N_edges = sum(c(O, LE, TE)))

###############
# create datasets
edge_dataset <- list()
for(i in 1:nrow(priority_data)){
  
  # first level prioritization
  keep_1 <- edge_data[which(edge_data[,priority_data$First[i]] == 1),]
  pos <- !which(names(keep_1) == priority_data$First[i]) == 2:4
  keep_1[,which(pos)+1] <- 0
  # second level prioritization
  keep_2 <- edge_data[which(edge_data[,priority_data$First[i]] == 0),] # remove 1st level
  keep_2 <- keep_2[which(keep_2$N_edges>1),] # select those with >1 edge
  keep_2 <- keep_2[which(keep_2[,priority_data$Second[i]] == 1),] # choose priority
  pos <- !which(names(keep_2) == priority_data$Second[i]) == 2:4
  keep_2[,which(pos)+1] <- 0
  # third level (keep if they do not overlap)
  keep_3 <- edge_data %>% filter(!spp %in% c(keep_1$spp,keep_2$spp))
  
  # Final
  edge_dataset[[i]] <- unique(rbind(keep_1, keep_2, keep_3))
  
}

```

### Rerun analysis - All edges model

```{r}

gam_sub <- pblapply(edge_dataset, function(x){
  
  edge_dataset_2go <- x
  
  LE_spp <- edge_dataset_2go$spp[which(edge_dataset_2go$LE==1)]
  O_spp <- edge_dataset_2go$spp[which(edge_dataset_2go$O==1)]
  TE_spp <- edge_dataset_2go$spp[which(edge_dataset_2go$TE==1)]
  
  # Fix param levels
  mydatatogo$Param <- relevel(mydatatogo$Param, ref = "O") 
  
  tmp <- glmer(SHIFT_abs ~ scale(vel_abs) * scale(GD) * Param +
                 LogNtempUnits + LogExtent + ContinuousGrain + PrAb + Quality +
                 (1 | Class),
               family = Gamma(link = "log"),
               data = mydatatogo %>%
                 filter((spp %in% LE_spp & Param == "LE") |
                          (spp %in% O_spp & Param == "O") |
                          (spp %in% TE_spp & Param == "TE")))
  
  tmp <- summary(tmp)
  tmp$coefficients
  
})

gam_sub

```

In most models the interaction between GD and velocity is significant!

------------------------------------------------------------------------
