---
title: "Adaptive potential models"
author: "Bioshifts group"
output:   
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
---


# Load libraries

```{r libs, message=FALSE}

# remotes::install_github("bbolker/broom.mixed")

rm(list=ls())

list.of.packages <- c(
  "ggplot2", "data.table", "glmmTMB", "lme4", "lmerTest","mgcv","itsadug","mgcViz","pbapply",
  # "lme4", "coefplot", "sjPlot", "sjmisc", "effects", "rgdal", "maptools", "terra", "MuMIn", "lsmeans", "GGally", "tidyterra", "DHARMa",
  "dplyr", "tidyr", "tidyverse","parallel","ggstats","optimx","broom.mixed","plotly"
) 

new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]

if(length(new.packages)) install.packages(new.packages)

sapply(list.of.packages, require, character.only = TRUE)

source(here::here("R/vis.gam.data_function.R"))

```


# Load data

```{r splist}

# Use only Fonseca
# mydataset <- read.csv(here::here("Data/gen_data_final_m5.csv"))
# mydataset <- mydataset %>% filter(Source == "Fonseca")

mydataset <- read.csv(here::here("Data/gen_data_final_m2_fonseca.csv"))

# how many data at each shift gradient
mydataset %>%
  group_by(Type) %>%
  tally

# how many data at each shift gradient / parameter
mydataset %>%
  group_by(Type,Param) %>%
  tally

summary(mydataset$velocity[which(mydataset$shift_vel_sign=="pospos")])
summary(mydataset$velocity[which(mydataset$shift_vel_sign=="negneg")])

summary(mydataset$SHIFT[which(mydataset$shift_vel_sign=="pospos")])
summary(mydataset$SHIFT[which(mydataset$shift_vel_sign=="negneg")])

summary(mydataset$velocity[which(mydataset$SHIFT>0)])
summary(mydataset$velocity[which(mydataset$SHIFT<0)])

```

# Data description

# Select only shifts in the same direction of velocity
```{r}

mydataset <- mydataset %>%
  filter(shift_vel_sign == "pospos" | shift_vel_sign == "negneg")
# 
# summary(mydataset$SHIFT[which(mydataset$shift_vel_sign=="pospos")])
# summary(mydataset$SHIFT[which(mydataset$shift_vel_sign=="negneg")])
# 
# summary(mydataset$velocity[which(mydataset$shift_vel_sign=="pospos")])
# summary(mydataset$velocity[which(mydataset$shift_vel_sign=="negneg")])


```

# Select groups with at least 5 observations and all three parameters
```{r}

n_groups = 5

type_groups_lat <- mydataset %>%
  filter(Type == "LAT") %>%
  group_by(Group,Param) %>%
  dplyr::summarise(N = length(Param)) %>%
  dplyr::filter(N>n_groups) %>%
  group_by(Group) %>%
  dplyr::summarise(N = length(Param)) %>%
  dplyr::filter(N>=3)

type_groups_ele <- mydataset %>%
  filter(Type == "ELE") %>%
  group_by(Group,Param) %>%
  dplyr::summarise(N = length(Param)) %>%
  dplyr::filter(N>n_groups) %>%
  group_by(Group) %>%
  dplyr::summarise(N = length(Param)) %>%
  dplyr::filter(N>=3)

mydataset <- mydataset %>%
  dplyr::filter((Type == "LAT" & Group %in% type_groups_lat$Group) | (Type == "ELE" & Group %in% type_groups_ele$Group))

```

# Remove outliers in shift

```{r message=FALSE}

# LAT
LAT_obs <- mydataset %>% 
  filter(Type == "LAT")

summary(LAT_obs$SHIFT_abs)
hist(LAT_obs$SHIFT_abs)

thr <- quantile(LAT_obs$SHIFT_abs,c(.01,0.99))
thr

rem <- which((LAT_obs$SHIFT_abs > thr[1]) & (LAT_obs$SHIFT_abs < thr[2]))
if(any(rem)){
  LAT_obs <- LAT_obs[rem,]
}
summary(LAT_obs$SHIFT_abs)
hist(LAT_obs$SHIFT_abs)

hist(sqrt(LAT_obs$SHIFT_abs))


# ELE
ELE_obs <- mydataset %>% 
  filter(Type == "ELE")

summary(ELE_obs$SHIFT_abs)
hist(ELE_obs$SHIFT_abs)

thr <- quantile(ELE_obs$SHIFT_abs,c(.01,0.99))
thr

rem <- which((ELE_obs$SHIFT_abs > thr[1]) & (ELE_obs$SHIFT_abs < thr[2]))
if(any(rem)){
  ELE_obs <- ELE_obs[rem,]
}
summary(ELE_obs$SHIFT_abs)
hist(ELE_obs$SHIFT_abs)

hist(sqrt(ELE_obs$SHIFT_abs))

# Combine
mydataset <- rbind(LAT_obs,ELE_obs)
hist(mydataset$SHIFT_abs)
hist(sqrt(mydataset$SHIFT_abs))


```

# Stats
```{r}

# N species
length(unique(mydataset$spp))

# N shifts
dim(mydataset)

```


# Latitude models

```{r message=FALSE}

mydatatogo <- mydataset  %>%
  dplyr::filter(Type == "LAT") %>%
  mutate(
    vel_raw = velocity,
    vel = scale(vel_raw),
    vel_abs = abs(vel_raw),
    vel_abs_scale = scale(vel_abs),
    vel_abs_log = log(vel_abs),
    vel_abs_log_scale = scale(vel_abs_log),
    vel_abs_sqrt = sqrt(vel_abs),
    vel_abs_sqrt_scale = scale(vel_abs_sqrt),
    
    SHIFT_raw = SHIFT,
    SHIFT = scale(SHIFT_raw),
    SHIFT_abs = abs(SHIFT_raw),
    SHIFT_abs_log = log(SHIFT_abs),
    SHIFT_abs_log_scale = scale(SHIFT_abs_log),
    
    SHIFT_cor_raw = SHIFT_cor,
    SHIFT_cor = scale(SHIFT_cor_raw),
    SHIFT_cor_abs = abs(SHIFT_cor_raw),
    SHIFT_cor_abs_log = log(SHIFT_cor_abs),
    SHIFT_cor_abs_log_scale = scale(SHIFT_abs_log),
    
    ID.area = scale(ID.area),
    DUR = scale(DUR),
    LogExtent = log(Extent),
    Lat = abs(Latitude),
    
    START = scale(START),
    Param = factor(Param),
    Group = factor(Group),
    spp = factor(spp),
    
    GD = Nucleotide_diversity,
    GD_sqrt = sqrt(GD),
    GD_scale = scale(GD),
    GD_sqrt_scale = scale(GD_sqrt)) %>%
  
  dplyr::select(
    GD, GD_sqrt, GD_scale, GD_sqrt_scale, Lat, TajimasD, overlap_SA, R2,
    SHIFT, SHIFT_abs, SHIFT_abs_log, SHIFT_raw, SHIFT_abs_log_scale,
    SHIFT_cor, SHIFT_cor_abs, SHIFT_cor_abs_log, SHIFT_cor_raw, SHIFT_abs_log_scale,
    vel, vel_abs, vel_abs_scale, vel_abs_log, vel_raw, vel_abs_log_scale, vel_abs_sqrt, vel_abs_sqrt_scale,
    # min_dist, dist_weight,
    START, DUR, ID.area, Nperiodes, LogNtempUnits, NtempUnits, Extent, LogExtent, ContinuousGrain, Quality, PrAb, ExtentF, AreaF,
    Param, Group, spp, Class, Order, Family, Genus, 
    ECO, Uncertainty_Parameter, Uncertainty_Distribution, Grain_size, Data, Sampling, Article_ID,
    # Marker,
    shift_vel_sign)

mydatatogo[,1:33] <- lapply(mydatatogo[,1:33], as.numeric)

mydatatogo <- data.frame(na.omit(mydatatogo))

mydatatogo$Article_ID <- as.factor(mydatatogo$Article_ID)
mydatatogo$Param <- as.factor(mydatatogo$Param)
mydatatogo$Class <- as.factor(mydatatogo$Class)
mydatatogo$Quality <- as.factor(mydatatogo$Quality)
mydatatogo$PrAb <- as.factor(mydatatogo$PrAb)
mydatatogo$ContinuousGrain <- as.numeric(mydatatogo$ContinuousGrain)

table(mydatatogo$Group)
# Bird   Fish Insect  Plant 
# 1273    228   1735    812 

table(mydatatogo$Class)
# Actinopterygii            Aves       Bryopsida Florideophyceae         Insecta      Liliopsida 
#             228            1273               2               1            1735             136 
# Magnoliopsida       Pinopsida  Polypodiopsida 
#             636              13              24 

mydatatogo <- filter(mydatatogo, Class %in% names(which(table(mydatatogo$Class) > 10)))
table(mydatatogo$Class)
# Actinopterygii           Aves        Insecta     Liliopsida  Magnoliopsida      Pinopsida Polypodiopsida 
#            228           1273           1735            136            636             13             24 

# N dataset
nrow(mydatatogo)
# nrow(mydatatogo %>% dplyr::filter(GD > 0))

# N species
length(unique(mydatatogo$spp))
# 1722

mydatatogo$Param <- relevel(mydatatogo$Param, ref = "O") 

table(mydatatogo$Param)

table(mydatatogo$ContinuousGrain)

table(mydatatogo$Quality)

table(mydatatogo$PrAb)

table(mydatatogo$ExtentF)

table(mydatatogo$Order)

```


## Histograms
### Shift
```{r message=FALSE}

hist(mydatatogo$SHIFT_abs)
hist(mydatatogo$SHIFT_abs_log)

```

### GD
```{r message=FALSE}
hist(mydatatogo$GD)
hist(mydatatogo$GD_sqrt)

### GD vs Group

ggplot(mydatatogo,
       aes(GD_sqrt, fill = Group, color = Group))+
  geom_density(alpha = .3)+
  theme_bw()

m <- aov(GD_sqrt ~ Group, mydatatogo)
summary(m)
TukeyHSD(m)

```

### Velocity
```{r message=FALSE}
hist(mydatatogo$vel_abs)
hist(mydatatogo$vel_abs_log)
```

We should use SHIFT_abs with a Gamma family or SHIFT_abs_log in a gaussian family.

## Scatter plots

### SHIFT_abs vs velocity_abs
```{r message=FALSE, fig.width=5, fig.height=5}
ggplot(mydatatogo,
       aes(x=vel_abs, y=SHIFT_abs))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()
```

### SHIFT_abs_log vs velocity_abs_log

```{r message=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=vel_abs_log, y=SHIFT_abs_log))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

```

### SHIFT_abs vs GD
```{r message=FALSE, fig.width=5, fig.height=5}
ggplot(mydatatogo,
       aes(x=GD, y=SHIFT_abs))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

ggplot(mydatatogo,
       aes(x=GD, y=SHIFT_abs))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

ggplot(mydatatogo,
       aes(x=GD, y=SHIFT_abs))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()+ 
  facet_wrap(.~Group, scales = "free")
```

### SHIFT_abs_log vs GD_sqrt

```{r message=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=GD_sqrt, y=SHIFT_abs_log))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

ggplot(mydatatogo,
       aes(x=GD_sqrt, y=SHIFT_abs_log))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()+ 
  facet_wrap(.~Class, scales = "free")


```

### SHIFT_abs vs Tajima's D
```{r message=FALSE, fig.width=5, fig.height=5}
ggplot(mydatatogo,
       aes(x=TajimasD, y=SHIFT_abs))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

ggplot(mydatatogo,
       aes(x=TajimasD, y=SHIFT_abs))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()+ 
  facet_wrap(.~Group, scales = "free")
```

### SHIFT_abs_log vs Tajima's D

```{r message=FALSE, fig.width=5, fig.height=5}
ggplot(mydatatogo,
       aes(x=TajimasD, y=SHIFT_abs_log))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

ggplot(mydatatogo,
       aes(x=TajimasD, y=SHIFT_abs_log))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()+ 
  facet_wrap(.~Group, scales = "free")


```





### Tajima's D vs GD_sqrt
```{r message=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=TajimasD, y=GD_sqrt))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

ggplot(mydatatogo,
       aes(x=TajimasD, y=GD_sqrt, color = Group))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

```

### Tajima's D vs Velocity
```{r message=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=vel_abs, y=TajimasD))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

```

### Tajima's D vs Velocity_abs_log
```{r message=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=vel_abs_log, y=TajimasD))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

```

### GD vs Velocity
```{r message=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=vel_abs, y=GD))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

```

### GD vs Velocity_abs_log
```{r message=FALSE, fig.width=5, fig.height=5}

ggplot(mydatatogo,
       aes(x=vel_abs_log, y=GD))+
  geom_point(alpha = .5, size = 2)+
  theme_bw()+
  geom_smooth()

```











## Modelling framework

Here we test the importance of intra-specific genetic diversity (GD) in modulating the response of species to climate exposure by shifting their ranges. 
GD is an important component of biodiversity because it provides adaptive potential (Frankham, 2005), facilitates species persistence (Frankham, 2005; Pereira et al., 2012), and provides clues about the historical demography of species (Avise, 2000). We expect that high GD would allow species to move faster to track climate change if its confers species' a higher potential to colonize new areas by increasing adaptive potential to the new community composition in newly colonized places. In contrast, high GD would allow species to stay in place if it confers species with higher plasticity or development of adaptations to tolerate new climate conditions.

Our modelling framework aims at exploring how species range shifts (km/year) (response variable) varies in function of climate change velocity (km/year) and GD (Nucleotide diversity). Specifically, our goal was to estimate whereas GD increases or decreases species range shift velocities in response to climate change velocity. Therefore, our model includes an interaction term of genetic diversity with climate change velocity.

Quaternary glaciation has long been recognized as a potential cause of spatial variation in genetic diversity (Hewitt, 2004). During the Quaternary, cyclical global cooling coupled with ice sheet advance and retreat promoted considerable demographic change and extinction (Hewitt, 2004; Nogués-Bravo et al., 2010). In order to control for potential confounding effect of Quaternary glaciation on genetic processes, we used the interaction term of GD with absolute latitude as a predictor variable. We did not include an interaction term of latitude with climate change velocity as we don't want to erase the fact that climate velocity is usually higher at high latitudes.

Our database comprise species across multiple taxonomic classes, with obvious differences in their ecologies and sensitivity to climate variability, which therefore causes observations to be non-independent from each other. We classify species in 4 major taxonomy groups (birds, plants, insects or fishes) in order to account for ecological differences among them (e.g., flight capacity would confer birds greater dispersal potential than plants, ectotherms and short lived like insects should be more sensitive to climate variability than endotherms like fishes). In order to control for taxonomy and ecological variability, we model Family and Genus nested into taxonomy group as a random intercept .

Our range shift estimates were gather from multiple literature sources. These shift estimates were collated using different methodological protocols, which cause non-independence among observations. Previous studies have shown that methodological aspects play an important role in explaining variability in species range shifts (Lenoir et al. 2020). To account for methodological differences in range shift estimates, we included range position nested within study ID as a random intercept. In addition to the random effect, we included fixed effects methodological variables representing number of temporal units, grain size (small, moderate or large), data type (abundance-based vs occurrence-based) and study area extent (log km). Removing these methodological variables from the model resulted in meaningless change in AIC, therefore we decided for keeping them out as to have a more parsimonious model. Finally, we also included a random effect to the study ID and range edge location (leading edge, trailing edge, centroid) from which range shifts were estimated. 

## GLM: GD vs velocity

### Gaussian vs Gamma model
```{r}

gau1 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (1|Class/Family), # Taxonomy
  data = mydatatogo)

# Works!
summary(gau1)

gam1 <- glmmTMB(
  SHIFT_abs ~ 
    vel_abs_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (1|Class/Family), # Taxonomy
  family = Gamma(link = "log"),
  data = mydatatogo)

# Works!
summary(gam1)

anova(gau1,gam1)

```


The Gaussian model have lower AIC. Use Gaussian model then!


#### Interpretation of results
Effect of GD is negative, meaning that species range shift less with high GD. Therefore, GD increases species' ability to stay in place.

Interaction of GD:Lat is positive, suggesting that at higher latitudes a higher GD leads to faster shifts.

Interaction of velocity and GD (vel_abs_log_scale:GD_sqrt_scale) is positive, indicating that as climate velocities increase species with higher GD move faster to track climate change. 

The positive effect of the interaction GD:velocity seems contradictory with the effect of GD alone. However, I believe they complement rather then contradict each other. The isolated negative effect of GD on range shifts indicates that overall GD increases species ability to stay in place. However, under scenarios of increasing climate change velocities, a higher GD will elevate species ability to track climate change.


```{r fig.height=5, fig.width=6}

plot(effects::effect("vel_abs_log_scale:GD_sqrt_scale", gau1))

```


### Trying alternative model structures
```{r}

# try with separate Article_ID and Param
gau2 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Article_ID) + # Methods II
    (1|Param) + # Methods II
    (1|Class/Family), # Taxonomy
  data = mydatatogo)

summary(gau2)

AIC(gau1, gau2)

#      df      AIC
# gau1 15 12578.25
# gau2 15 12620.09

# gau1 is better

```


```{r}

# try random slope 
gau3 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (vel_abs_log_scale : GD_sqrt_scale|Class/Family), # Taxonomy
  data = mydatatogo)
# Warning: Model convergence problem; singular convergence (7). See vignette('troubleshooting'), help('diagnose')
diagnose(gau3)
# random slope on vel_abs_log_scale:GD_sqrt_scale|Class is not helping
summary(gau3) 

gau3 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (vel_abs_log_scale : GD_sqrt_scale|Group) + # Taxonomy
    (1|Class/Family),
  data = mydatatogo)
# Warning: Model convergence problem; non-positive-definite Hessian matrix. See vignette('troubleshooting')
diagnose(gau3)

summary(gau3) # random slope on vel_abs_log_scale:GD_sqrt_scale|Group is not helping


gau3 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (vel_abs_log_scale : GD_sqrt_scale|Group) + # Taxonomy
    (1|Order/Family),
  data = mydatatogo)
# Warning: Model convergence problem; non-positive-definite Hessian matrix. See vignette('troubleshooting')

summary(gau3) # random slope on vel_abs_log_scale:GD_sqrt_scale|Group is not helping

gau3 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (vel_abs_log_scale : GD_sqrt_scale|Class), # Taxonomy
  data = mydatatogo)
# Doesn't converges but fits...
# random slope on vel_abs_log_scale:GD_sqrt_scale|Class is not helping
# the interaction of vel_abs_log_scale:GD_sqrt_scale becomes is significant!
summary(gau3) 

AIC(gau1,gau3)
# gau1 is better

```


We have 2 candidate models:
*The simple:* (gau1)
SHIFT_abs_log ~ vel_abs_log_scale * GD_sqrt_scale + GD_sqrt_scale:Lat +  
LogExtent + ContinuousGrain + Quality + PrAb + (1 | Param:Article_ID) +  
(1 | Order/Family)

*The random slope:* (gau3)
SHIFT_abs_log ~ vel_abs_log_scale * GD_sqrt_scale + GD_sqrt_scale:Lat +  
LogExtent + ContinuousGrain + Quality + PrAb + (1 | Param:Article_ID) +  
(vel_abs_log_scale:GD_sqrt_scale | Class)



```{r}
summary(gau1)
summary(gau3)
```


```{r fig.height=5, fig.width=6}

plot(effects::effect("vel_abs_log_scale:GD_sqrt_scale", gau1), main = "Model 1")
plot(effects::effect("vel_abs_log_scale:GD_sqrt_scale", gau3), main = "Model 3")

```


## GLM: Tajima's vs velocity

We went further and use a different metric to investigate the potential modulating effect of genetic processes on range shifts. 
Tajima's D is widely used to detect departure from neutrality that potentially result from population size change over time  (Tajima, 1989). Tajima’s D is composed of two other metrics: the mean pairwise difference (π) and the number (S) of segregating sites, both representing an estimator of θ (theta). While the first metric measures the observed genetic variation in the sequences, the second metric quantifies the expected amount of genetic variation under the null model. Thus, under a scenario of constant population size, the observed amount of genetic variation (π) should be approximately equal to the expected genetic variation (S). For example, population expansions increase the number of rare variants (i.e., a polymorphic site present in only one—singletons—or in very few individuals) in the population, which leads to an increase in the expected value of theta. Conversely, rare polymorphisms lead to a lower value of π. The distortion between π and S, which represents a departure from mutation-drift equilibrium, is captured by Tajima’s D. Negative values are interpreted as an evidence of population expansion and positive values as population bottleneck. Note that while some authors have argued that significant results could result from positive or balancing selection, respectively, the demographic causes are the most common interpretation in phylogeographic studies and likely to be the most relevant for a global comparison (Fonseca et al. 2023). 

Here, we keep the same model structure as before, but include Tajima's D as an interaction term with climate velocity.

```{r height=10, fig.width=10}

gau11 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * TajimasD +# Ecology
    TajimasD:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (1|Class/Family), # Taxonomy
  data = mydatatogo)

summary(gau11)

gau13 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * TajimasD + # Ecology
    TajimasD:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (vel_abs_log_scale : TajimasD|Class), # Taxonomy
  data = mydatatogo)
# convergence issue
summary(gau13) 

gau13 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * TajimasD + # Ecology
    TajimasD:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (vel_abs_log_scale : TajimasD|Group), # Taxonomy
  data = mydatatogo)
# Works
# but not significant vel_abs_log_scale:TajimasD
summary(gau13) 

gau13 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * TajimasD + # Ecology
    TajimasD:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (vel_abs_log_scale : TajimasD|Class:Family), # Taxonomy
  data = mydatatogo)
# Works
summary(gau13) 
# Group:Family vel_abs_log_scale:TajimasD explains very little


AIC(gau11,gau13)
# gau11 is better

```

### Interpretation of results 
Range shifts are positively associated with climate velocity, suggesting shift magnitudes increase with climate exposure. Range shifts are negatively correlated with Tajima's D, suggesting species that show signature of population contraction shift less. Moreover, the interaction term indicates range shifts are positively associated with climate velocity when Tajima's D is positive (evidence of population contraction). This indicates that under populations contraction, species tend to track climate change more closely (otherwise they go extinct). In contrast, when populations expand species are more likely to remain in place and don't move to track climate change. 


## GAMM: GD vs velocity

```{r }
# first we contrast models with a thin plate regression spline (default smoother in GAM) with models with a tensor smooth product

# GD_gamm1 <- gamm(
#   SHIFT_abs_log ~
#     GD_sqrt_scale + vel_abs_log_scale + GD_sqrt_scale:Lat +
#     s(GD_sqrt_scale, vel_abs_log_scale) + # interaction between smoothed variables
#     LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
#   random = list(
#     Param=~1, # random effect of Param
#     Article_ID=~1, # random effect of Study
#     Class=~1), # random effect of Class
#   data = mydatatogo)
# singularity issues

GD_gamm1 <- gamm(
  SHIFT_abs_log ~ 
    s(GD_sqrt_scale, vel_abs_log_scale) + # interaction between smoothed variables
    LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
  random = list(
    Param=~1, # random effect of Param
    Article_ID=~1, # random effect of Study
    Class=~1), # random effect of Class
  data = mydatatogo)

summary(GD_gamm1$gam)

GD_gamm2 <- gamm(
  SHIFT_abs_log ~ 
    ti(GD_sqrt_scale, vel_abs_log_scale) + # interaction between smoothed variables (using a tensor smoother)
    LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
  random = list(
    Param=~1, # random effect of Param
    Article_ID=~1, # random effect of Study
    Class=~1), # random effect of Class
  data = mydatatogo)

summary(GD_gamm2$gam)

anova(GD_gamm1$lme, GD_gamm2$lme)

```

Models 1 and 2 are identical based on AIC. I will keep the tensor smoother from model2 as it leads to higher r-squared and <P-values for the interaction term...

In addition, I check if including the GD and velocity in the model helps...
```{r }

GD_gamm3 <- gamm(
  SHIFT_abs_log ~ 
    GD_sqrt_scale + vel_abs_log_scale + GD_sqrt_scale:Lat +
    ti(GD_sqrt_scale, vel_abs_log_scale) + # interaction between smoothed variables (using a tensor smoother)
    LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
  random = list(
    Param=~1, # random effect of Param
    Article_ID=~1, # random effect of Study
    Class=~1), # random effect of Class
  data = mydatatogo)

summary(GD_gamm3$gam)

anova(GD_gamm2$lme, GD_gamm3$lme)


```

There is no difference between model 2 and 3. I will keep model 3 as it leads to higher r-squared and <P-values for the interaction term...

```{r fig.height=5, fig.width=5}

dataplot <- vis.gam.data(GD_gamm3$gam, 
                         view = c("GD_sqrt_scale","vel_abs_log_scale"), 
                         n.grid = 10, 
                         theta = 45)
GD <- dataplot[[2]]
Velocity <- dataplot[[1]]
Response = dataplot[[3]]

plot_ly(x = ~Velocity, y = ~GD, z = ~Response, type = "surface") %>% 
  layout(scene = list(
    aspectmode='cube',
    yaxis = list(autorange = "reversed")))

```

Test if including a random slope helps...

```{r height=5, fig.width=5}

GD_gamm4 <- gamm(
  SHIFT_abs_log ~ 
    GD_sqrt_scale + vel_abs_log_scale + GD_sqrt_scale:Lat +
    ti(GD_sqrt_scale, vel_abs_log_scale) + # interaction between smoothed variables (using a tensor smoother)
    LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
  random = list(
    Param=~1, # random effect of Param
    Article_ID=~1, # random effect of Study
    Class=~GD_sqrt_scale), # random slope of GD on Class
  data = mydatatogo)

summary(GD_gamm4$gam)

anova(GD_gamm3$lme, GD_gamm4$lme)

```

No!


## GAMM: Tajima's vs velocity

```{r height=5, fig.width=5}

# Taj_gamm1 <- gamm(
#   SHIFT_abs_log ~ 
#     TajimasD + vel_abs_log_scale + TajimasD:Lat +
#     s(TajimasD, vel_abs_log_scale) + # interaction between smoothed variables (using a tensor smoother)
#     LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
#   random = list(
#     Param=~1, # random effect of Param
#     Article_ID=~1,
#     Class=~1), # random effect of Study
#   data = mydatatogo)
# singularity issue

Taj_gamm2 <- gamm(
  SHIFT_abs_log ~ 
    TajimasD + vel_abs_log_scale + TajimasD:Lat +
    ti(TajimasD, vel_abs_log_scale) + # interaction between smoothed variables (using a tensor smoother)
    LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
  random = list(
    Param=~1, # random effect of Param
    Article_ID=~1,
    Class=~1), # random effect of Study
  data = mydatatogo)

summary(Taj_gamm2$gam)

```




```{r height=10, fig.width=10}
dataplot <- vis.gam.data(Taj_gamm2$gam, 
                         view = c("TajimasD","vel_abs_log_scale"), 
                         n.grid = 10, 
                         theta = 45)
TajimasD <- dataplot[[2]]
Velocity <- dataplot[[1]]
Response = dataplot[[3]]

plot_ly(x = ~Velocity, y = ~TajimasD, z = ~Response, type = "surface") %>% 
  layout(scene = list(
    aspectmode='cube',
    yaxis = list(autorange = "reversed")))

```

# Elevation models

```{r message=FALSE}

mydatatogo <- mydataset  %>%
  dplyr::filter(Type == "ELE") %>%
  mutate(
    vel_raw = velocity,
    vel = scale(vel_raw),
    vel_abs = abs(vel_raw),
    vel_abs_scale = scale(vel_abs),
    vel_abs_log = log(vel_abs),
    vel_abs_log_scale = scale(vel_abs_log),
    vel_abs_sqrt = sqrt(vel_abs),
    vel_abs_sqrt_scale = scale(vel_abs_sqrt),
    
    SHIFT_raw = SHIFT,
    SHIFT = scale(SHIFT_raw),
    SHIFT_abs = abs(SHIFT_raw),
    SHIFT_abs_log = log(SHIFT_abs),
    SHIFT_abs_log_scale = scale(SHIFT_abs_log),
    
    SHIFT_cor_raw = SHIFT_cor,
    SHIFT_cor = scale(SHIFT_cor_raw),
    SHIFT_cor_abs = abs(SHIFT_cor_raw),
    SHIFT_cor_abs_log = log(SHIFT_cor_abs),
    SHIFT_cor_abs_log_scale = scale(SHIFT_abs_log),
    
    ID.area = scale(ID.area),
    DUR = scale(DUR),
    LogExtent = log(Extent),
    Lat = abs(Latitude),
    
    START = scale(START),
    Param = factor(Param),
    Group = factor(Group),
    spp = factor(spp),
    
    GD = Nucleotide_diversity,
    GD_sqrt = sqrt(GD),
    GD_scale = scale(Nucleotide_diversity),
    GD_sqrt_scale = scale(sqrt(GD))) %>%
  
  dplyr::select(
    GD, GD_sqrt, GD_scale, GD_sqrt_scale, Lat, TajimasD, overlap_SA, R2,
    SHIFT, SHIFT_abs, SHIFT_abs_log, SHIFT_raw, SHIFT_abs_log_scale,
    SHIFT_cor, SHIFT_cor_abs, SHIFT_cor_abs_log, SHIFT_cor_raw, SHIFT_abs_log_scale,
    vel, vel_abs, vel_abs_scale, vel_abs_log, vel_raw, vel_abs_log_scale, vel_abs_sqrt, vel_abs_sqrt_scale,
    # min_dist, dist_weight,
    START, DUR, ID.area, Nperiodes, LogNtempUnits, NtempUnits, Extent, LogExtent, ContinuousGrain, Quality, PrAb, ExtentF, AreaF,
    Param, Group, spp, Class, Order, Family, Genus, 
    ECO, Uncertainty_Parameter, Uncertainty_Distribution, Grain_size, Data, Sampling, Article_ID,
    # Marker,
    shift_vel_sign)

mydatatogo[,1:33] <- lapply(mydatatogo[,1:33], as.numeric)

mydatatogo <- data.frame(na.omit(mydatatogo))

# N dataset
nrow(mydatatogo)
nrow(mydatatogo %>% dplyr::filter(GD > 0))

# N species
length(unique(mydatatogo$spp))
# 1725

table(mydatatogo$Group)
# Bird   Fish Insect  Plant 
# 1273    228   1735    812 
mydatatogo$Param <- relevel(mydatatogo$Param, ref = "O") 

table(mydatatogo$Param)

table(mydatatogo$ContinuousGrain)

table(mydatatogo$Quality)

table(mydatatogo$PrAb)

table(mydatatogo$ExtentF)

table(mydatatogo$Order)

```

## GLM: GD vs velocity
### Gaussian vs Gamma model
```{r}

gau1 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (1|Class/Family), # Taxonomy
  data = mydatatogo)

# Works!
summary(gau1)

gam1 <- glmmTMB(
  SHIFT_abs ~ 
    vel_abs_scale * GD_sqrt_scale + # Ecology
    GD_sqrt_scale:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (1|Class/Family), # Taxonomy
  family = Gamma(link = "log"),
  data = mydatatogo)

# Works!
summary(gam1)

anova(gau1,gam1)

```

No significant effects

## GLM: Tajima's vs velocity
```{r}

gau11 <- glmmTMB(
  SHIFT_abs_log ~ 
    vel_abs_log_scale * TajimasD +# Ecology
    TajimasD:Lat + # LGM effect on GD
    LogExtent + ContinuousGrain + Quality + PrAb + # Methods
    (1|Param/Article_ID) + # Methods II
    (1|Class/Family), # Taxonomy
  data = mydatatogo)

summary(gau11)


```


No significant effects

## GAMM: GD vs velocity

```{r height=5, fig.width=5}

GD_gamm3 <- gamm(
  SHIFT_abs_log ~ 
    GD_sqrt_scale + vel_abs_log_scale + GD_sqrt_scale:Lat +
    ti(GD_sqrt_scale, vel_abs_log_scale) + # interaction between smoothed variables (using a tensor smoother)
    LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
  random = list(
    Param=~1, # random effect of Param
    Article_ID=~1, # random effect of Study
    Class=~1), # random effect of Class
  data = mydatatogo)

summary(GD_gamm3$gam)


```


No significant effects


## GAMM: Tajima's vs velocity

```{r height=5, fig.width=5}

Taj_gamm2 <- gamm(
  SHIFT_abs_log ~ 
    TajimasD + vel_abs_log_scale + TajimasD:Lat +
    ti(TajimasD, vel_abs_log_scale) + # interaction between smoothed variables (using a tensor smoother)
    LogExtent + ContinuousGrain + Quality + PrAb, # fixed effects of methods
  random = list(
    Param=~1, # random effect of Param
    Article_ID=~1,
    Class=~1), # random effect of Study
  data = mydatatogo)

summary(Taj_gamm2$gam) 

```

No significant effects




